br/pkg/checkpoint/checkpoint.go

```diff
@@ -305,12 +307,79 @@ func (r *CheckpointRunner[K, V]) setLock(ctx context.Context, errCh chan error)
 	return nil
 }
 
+type flusher[K KeyType, V ValueType] struct {
+	incompleteMetas     []map[K]*RangeGroup[K, V]
+	incompleteChecksums []ChecksumItems
+}
+
+func newFlusher[K KeyType, V ValueType]() *flusher[K, V] {
+	return &flusher[K, V]{
+		incompleteMetas:     make([]map[K]*RangeGroup[K, V], 0),
+		incompleteChecksums: make([]ChecksumItems, 0),
+	}
+}
+
+func (f *flusher[K, V]) doFlush(ctx context.Context, r *CheckpointRunner[K, V], meta map[K]*RangeGroup[K, V]) {
+	if err := r.doFlush(ctx, meta); err != nil {
+		log.Warn("failed to flush checkpoint data", zap.Error(err))
+		f.incompleteMetas = append(f.incompleteMetas, meta)
+	}
+}
+
+func (f *flusher[K, V]) doChecksumFlush(ctx context.Context, r *CheckpointRunner[K, V], checksums ChecksumItems) {
+	if err := r.doChecksumFlush(ctx, checksums); err != nil {
+		log.Warn("failed to flush checkpoint checksum", zap.Error(err))
+		f.incompleteChecksums = append(f.incompleteChecksums, checksums)
+	}
+}
+
+func (f *flusher[K, V]) flushOneIncomplete(ctx context.Context, r *CheckpointRunner[K, V]) {
+	if len(f.incompleteMetas) > 0 {
+		lastIdx := len(f.incompleteMetas) - 1
+		if err := r.doFlush(ctx, f.incompleteMetas[lastIdx]); err != nil {
+			log.Warn("failed to flush checkpoint data", zap.Error(err))
+			return
+		}
+		f.incompleteMetas = f.incompleteMetas[:lastIdx]
+	} else if len(f.incompleteChecksums) > 0 {
+		lastIdx := len(f.incompleteChecksums) - 1
+		if err := r.doChecksumFlush(ctx, f.incompleteChecksums[lastIdx]); err != nil {
+			log.Warn("failed to flush checkpoint data", zap.Error(err))
```

Comment:
```suggestion
			log.Warn("failed to flush checkpoint checksum", zap.Error(err))
```

it's better to have some other log, because we only doing flushOneIncomplete in retry.

Reply:
done